capirca/lib/fortigate.py

```diff
@@ -0,0 +1,937 @@
+
+    """
+    addr_name = addr.with_prefixlen
+    address = []
+    address += [f'{_SP} edit "{addr_name}"']
+    if verbose:
+      addr_comment = _DEFAULT_COMMENT
+      if len(addr.text):
+        addr_comment += " (" + addr.text + ")"
+      address += [f'{_SP * 2} set comment "{self.fix_comment_length(addr_comment)}"']
+    if not isinstance(addr, nacaddr.IPv6):
+      address += [f'{_SP * 2} set subnet {addr_name}']
+    if isinstance(addr, nacaddr.IPv6):
+      address += [f'{_SP * 2} set ip6 {addr_name}']
+    address += [_SP + 'next']
+
+    return address
+
+  def get_fw_addrgrps(self, ip_v=4):
+    """Returns the collected address groups."""
+    addrgrps = []
+    for addrgrp_name, addrgrp_item in self._dict_addrgrps.items():
+      if addrgrp_item[0] == ip_v:
+        address = addrgrp_item[1]
+        exclude_address = addrgrp_item[2]
+        addrgrps += [f'{_SP} edit "{addrgrp_name}"']
+        if self.verbose:
+          addrgrps += [f'{_SP * 2} set comment "{_DEFAULT_COMMENT}"']
+        if address:
+          addrgrps += [f'{_SP * 2} set member {" ".join(f"\"{v}\"" for v in address)}']
+        else:
+          addrgrps += [_SP * 2 + 'set member "all"']
+        if exclude_address:
+          addrgrps += [_SP * 2 + 'set exclude enable']
+          addrgrps += [f'{_SP * 2} set exclude-member {" ".join(f"\"{v}\"" for v in exclude_address)}']
+        addrgrps += [_SP + 'next']
+
+    return addrgrps
+
+  def get_port_range(self, port):
+    """
+    Returns the port range as a string.
+
+    If the input port is a single integer, returns that value.
+    If the input port is a list of two integers, returns a range string (e.g. "1-10").
+
+    :param port: The port number or list of port numbers
+    :return: A string representing the port range
+    """
+    port_range = ''
+    if len(port) == 1:
+        port_range = str(port[0])
+    elif len(port) > 1:
+        port_range = str(port[0])
+        if port[0] != port[1]:
+            port_range = str(min(port[0], port[1])) + '-' + str(max(port[0], port[1]))
+
+    return port_range
+
+  def get_fw_services(self):
+    """Returns the collected services."""
+    fw_services = []
+    for service_name in sorted(self._dict_services.keys()):
+      fw_services += [_SP + 'edit ' + service_name]
+      if self.verbose:
+        fw_services += [f'{_SP * 2} set comment {_DEFAULT_COMMENT}']
+      for service_item in self._dict_services[service_name]:
+        fw_services += [_SP * 2 + service_item]
+      fw_services += [_SP + 'next']
+
+    return fw_services
+
+  def get_fw_svcgrps(self):
+    """Returns the collected service groups."""
+    svcgrps = []
+    for svcgrp, value in self._dict_svcgrps.items():
+      svcgrps += [f'{_SP} edit {svcgrp}']
+      if self.verbose:
+        svcgrps += [f'{_SP * 2} set comment {_DEFAULT_COMMENT}']
+      svcgrps += [_SP * 2 + value]
+      svcgrps += [_SP + 'next']
+
+    return svcgrps
+
+  def get_fw_schedules(self):
```

Comment:
[no-op] This is an interesting way to address expirations - most other platforms just don't render expired terms.